lib/ruby_lsp/extension.rb

```diff
@@ -0,0 +1,53 @@
+# typed: strict
+# frozen_string_literal: true
+
+module RubyLsp
+  class Extension
+    class << self
+      extend T::Sig
+      extend T::Helpers
+
+      abstract!
+
+      sig { params(child_class: T.class_of(Extension)).void }
+      def inherited(child_class)
+        extensions << child_class
+        super
+      end
+
+      sig { returns(T::Array[T.class_of(Extension)]) }
+      def extensions
+        @extensions ||= T.let([], T.nilable(T::Array[T.class_of(Extension)]))
+      end
+
+      sig { returns(T::Array[StandardError]) }
+      def load_extensions
+        # Require all extensions entry points, which should be placed under
+        # `some_gem/lib/ruby_lsp/your_gem_name/extension.rb`
+        errors = ::Gem.find_files("ruby_lsp/*/extension.rb").filter_map do |extension|
+          require File.expand_path(extension)
+          nil
+        rescue => e
+          e
+        end
+
+        # Activate each one of the discovered extensions. If any problems occur in the extensions, we don't want to
+        # fail to boot the server
+        extensions.each do |extension|
+          extension.activate
+        rescue => e
+          errors << e
+        end
+
+        errors
```

Comment:
I wonder if it'll be better to not just return errors but also other info. For example, `errors` is actually an array of hashes like: 

```
{
  extension_file: # for error during require
  extension_class: # for error during activation
  error: e
}
```

Otherwise, if there are multiple extensions and an error like `NoMethodError` on `nil` is being raised, just seeing `Exception#message` won't be too helpful. At least we should let the user know which extension caused that?

Reply:
Along the same lines of what @st0012 mentioned, returning an array of `StandardError` instance feels not quite right here. Maybe if they could be encapsulated in something like an array of `ExtensionLoadResult`.

Reply:
Okay, I changed the way extensions are registered in the last commit to allow keeping track of names and better encapsulate errors. Let me know what you think.